src/relay/op/type_relations.cc

```diff
@@ -0,0 +1,137 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file type_relations.cc
+ * \brief A set of utilities and common functionality
+ * for type relations.
+ */
+#include <tvm/relay/expr.h>
+#include <tvm/relay/logging.h>
+#include <tvm/relay/op.h>
+#include "../pass/incomplete_type.h"
+
+namespace tvm {
+namespace relay {
+
+TensorType as_ttype(const Type& t) {
+  if (auto tt_node = t.as<TensorTypeNode>()) {
+    return GetRef<TensorType>(tt_node);
+  } else {
+    return TensorType(nullptr);
+  }
+}
+
+// TODO(@jroesch) what size value do we extract?
+int to_int(const tvm::Expr& e) {
+  CHECK(e.defined());
+  auto imm = e.as<tvm::ir::IntImm>();
+  CHECK(imm) << "TYPE: " << imm << imm->type << std::endl;
+  return imm->value;
+}
+
+Array<Type> IdentityRel(const Array<Type>& types, int num_args) {
+  CHECK_EQ(types.size(), 2);
+  auto t1 = as_ttype(types[0]);
+  if (t1 && types[1].as<IncompleteTypeNode>()) {
```

Comment:
Also support backward inference, types[1] is known and types[0] is note. We need to have a way to notify type checking error(when both are known but inconsistent). Maybe thrown an error with good error message.

In NNVM, there is NNVM_SHAPE_ASSIGN macro that does that